lib_eio/unix/thread_pool.ml

```diff
@@ -0,0 +1,216 @@
+
+   Note: type ['a] will be [Mailbox.t], defined below. *)
+type 'a t = {
+  mutable initialized: bool; (* Have we setup the [Domain.at_exit] callback? *)
+  threads: 'a array; (* An array of [Mailbox.t] *)
+  available: Bitfield.t; (* For each thread, is it ready to receive work? *)
+  mutable terminating: bool; (* Is the domain exiting? *)
+}
+
+module Mailbox = struct
+  (** Mailbox with blocking OS semaphore.
+     Each [Mailbox.t] controls one worker systhread.
+     After completing each job, a thread will be kept for a short period
+     of time in case more work is immediately available.
+     Otherwise, the thread exits. *)
+  type t = {
+    lock: Timed_semaphore.t; (* A Unix [pthread_cond_t] *)
+    mutable cell: job;
+    mutable i: int; (* The index of this thread in the threadpool arrays *)
+  }
+
+  (* A worker systhread does not start with an assigned index.
+     [i] cannot be an option because writing a Some to this field would allocate. *)
+  let create () = { lock = Timed_semaphore.make false; cell = New; i = -1 }
+
+  let dummy = create ()
+
+  let put mbox x =
+    mbox.cell <- x;
+    Timed_semaphore.release mbox.lock
+
+  (* [@poll error] ensures this function is atomic within systhreads of a single domain.
+     Returns [true] if the thread should exit. *)
+  let[@poll error] handle_timedout pool = function
+  | { i = -1; _} ->
+    (* This thread was never placed in the pool, there is nothing to clean up *)
+    true
+  | { i; _ } ->
+    match Bitfield.get pool.available i with
+    | true ->
+      (* Cleanup and exit *)
+      Bitfield.set pool.available i false;
+      true
+    | false ->
+      (* A thread switch happened right before the start of this function.
+         [Timed_semaphore.acquire_with_timeout] timed out, but a new job came in
+         for this thread before we could stop the thread, so execute the new job. *)
+      false
+
+  let rec take pool mbox =
+    if Timed_semaphore.acquire_with_timeout mbox.lock max_wait_seconds
+    then mbox.cell
+    else (
+      (* Semaphore timed out *)
+      if handle_timedout pool mbox
+      then Exit
+      else take pool mbox
+    )
+end
+
+let create () =
+  {
+    initialized = false;
+    threads = Array.make max_standby_systhreads_per_domain Mailbox.dummy;
+    available = Bitfield.create ();
+    terminating = false;
+  }
+
+(* This function is necessary in order to immediately terminate the systhreads
+   on stand-by, without having to wait for them to shut down automatically.
+   Without this termination mechanism, domains that have executed a call on
+   a systhread in the last [max_wait_seconds] seconds would have to wait for
+   their timeout to occur, introducing long unwanted pauses. *)
+let terminate pool =
+  pool.terminating <- true;
+  if not (Bitfield.is_empty pool.available) then
+    for i = 0 to max_standby_systhreads_per_domain - 1 do
+      if Bitfield.get pool.available i
+      then Mailbox.put pool.threads.(i) Exit
```

Comment:
Hmm... It is usually prudent to join with any domains, threads, or fibers.  Granted there shouldn't be much that could go wrong here.